hw/nfc-nci.c

```diff
@@ -713,31 +725,45 @@ init_process_oid_rf_deactivate_cmd(const union nci_packet* cmd,
             rfst = NUMBER_OF_NFC_RFSTS;
     }
 
+    /* reset state */
+
+    nfc->active_re = NULL;
+    nfc->active_rf = NULL;
+
+    for (i = 0; i < sizeof(nfc_res)/sizeof(nfc_res[0]); ++i) {
+        nfc_res[i].id = 0;
+    }
+
+    rfst = nfc_rf_state_transition(&nfc->rf_state, bits, rfst);
+    assert(rfst != NUMBER_OF_NFC_RFSTS);
+}
+
+static size_t
+init_process_oid_rf_deactivate_cmd(const union nci_packet* cmd,
+                                   struct nfc_device* nfc,
+                                   union nci_packet* rsp,
+                                   struct nfc_delivery_cb* cb)
+{
+    const struct nci_rf_deactivate_cmd *payload;
+    int send_ntf = 1;
+
+    payload = (struct nci_rf_deactivate_cmd*)cmd->control.payload;
+
     // According to NCI Spec Figure 10.
     switch (nfc->rf_state) {
         case NFC_RFST_DISCOVERY:
         case NFC_RFST_W4_ALL_DISCOVERIES:
         case NFC_RFST_W4_HOST_SELECT:
         case NFC_RFST_LISTEN_SLEEP:
-            if (payload->type == NCI_RF_DEACT_IDLE_MODE)
+            if (payload->type == NCI_RF_DEACT_IDLE_MODE) {
+                nfc_rf_state_deactivate_transition(payload, nfc);
                 send_ntf = false;
```

Comment:
If i understand correctly, what you are trying to do here is move state transition to deactivate_ntf.
So it will be better move |nfc_rf_state_deactivate_transition| to else of |if(send_ntf)|  check